uffd: uffd_open caller prints info/error instead

angieenee opened the pull request:
When uffd_open is called from kerndat_uffd, userfaultfd failiure is not
considered an error, so the goal is to suppress the error message --
instead, the goal of this change is to print this message as info.

Initially, uffd_open handles userfaultfd fails with errors only, but this
implementation distinguishes whether to report as an error or info
depending on where uffd_open was called.

Signed-off-by: Angie Ni <avtni@google.com>

angieenee commented on criu/uffd.c:
```diff
@@ -319,7 +316,9 @@ int setup_uffd(int pid, struct task_restore_args *task_args)
 	if (task_args->uffd < 0) {
-		pr_perror("Unable to open an userfaultfd descriptor");
+		if (task_args->uffd == -errno)
+			pr_error("Lazy pages are not available");
```
oh no

kashyapp commented:
Looks like something went wrong in the commit list.
There's a number of upstream commits.


A useful command to know is show-branch, it shows the list of commits of the two branches up until their intersection point.
git show-branch 0013 0013-wip

As Michal suggested, you could rebase to fix this

    git checkout 0013-wip
    git rebase -i 0013 
    # in the editor, drop all the upstream commits (replace pick w drop), retain your own

kashyapp commented on criu/uffd.c:
```diff
@@ -313,9 +314,11 @@ int setup_uffd(int pid, struct task_restore_args *task_args)
+	task_args->uffd = uffd_open(O_CLOEXEC | O_NONBLOCK, &features, false);
 	if (task_args->uffd < 0) {
-		pr_perror("Unable to open an userfaultfd descriptor");
+		if (task_args->uffd == -errno)
```
I'm not sure I understand ths, why compare to `-errno`?

kashyapp replied:
In any case, I think what rppt suggested in https://github.com/checkpoint-restore/criu/pull/1094#pullrequestreview-428418065 is this

There are only 2 callers to uffd_open. Instead of adding an additional parameter to control the log level there, their suggestion is to move the logging responsibility to the caller itself.

angieenee replied:
I see. I also noticed I made a critical error by not providing a different return value if log_level is true. My intent was to have uffd_open() print info if log_level was set to true, else return -errno so the caller can print the error. However, I did not execute this properly. 

I think you are right that I should have the callers be responsible for printing the error/info.

angieenee replied:
I use -errno because there were different areas in uffd_open where something could have gone wrong. To make the message more specific, I wanted -errno to indicate that there were no available lazy pages. My intent with using -errno came from rppt's suggestion of changing the wording of the errors. 

However, I think you are right. I will make the callers responsible for handling whether to print as info or error. Thanks, @kashyapp !

mclapinski commented on criu/kerndat.c:
```diff
@@ -912,7 +912,7 @@ static int kerndat_uffd(void)
 			return 0;
 
-		pr_err("Lazy pages are not available\n");
+		pr_info("Lazy pages are not available: %s\n", strerror(errno));
```
revert this

mclapinski commented on criu/uffd.c:
```diff
@@ -317,9 +311,11 @@ int setup_uffd(int pid, struct task_restore_args *task_args)
-		pr_perror("Unable to open an userfaultfd descriptor");
+		if (task_args->uffd == -errno)
+			pr_perror("Lazy pages are not available");
+		else pr_perror("Unable to open an userfaultfd descriptor");
```
i'd revert those lines

mclapinski commented on criu/uffd.c:
```diff
@@ -262,20 +262,14 @@ static int uffd_api_ioctl(void *arg, int fd, pid_t pid)
-		} else {
-			pr_perror("Lazy pages are not available");
-		}
+	if (uffd == -1)
```
add pr_info in this if

Commit:
uffd: uffd_open prints info, caller prints error

When uffd_open is called from kerndat_uffd, userfaultfd failiure is not
considered an error, so the goal is to suppress the error message --
instead, the goal of this change is to print this message as info.

Initially, uffd_open handles userfaultfd fails with errors only, but this
implementation distinguishes whether to report as an error or info
depending on where uffd_open was called.

Signed-off-by: Angie Ni <avtni@google.com>
